### PR TITLE
JDK-8257828: SafeFetch may crash if invoked in non-JavaThreads

### PR DESCRIPTION
**Pull request by tstuefe**
In our primary hotspot signal handlers, SafeFetch handling is limited to JavaThread objects:

```
  JavaThread* thread = NULL;
...
  if(t->is_Java_thread()) {
    thread = (JavaThread*)t;
  }
...
  if (info != NULL && uc != NULL && thread != NULL) {
    pc = (address) os::Linux::ucontext_get_pc(uc);
    if (StubRoutines::is_safefetch_fault(pc)) {
```

As a result of this, using SafeFetch may crash non-JavaThreads if the location is invalid. E.g. using SafeFetch inside a VMOperation may crash the VM.

This is unfortunate since SafeFetch is used for os::is_readable_pointer() which explicitly promises to not crash. It is used e.g. in os::print_hex_dump(). There is also no reason why SafeFetch would not work for non-JavaThreads. In fact, SafeFetch handling for the secondary signal handler works just fine for all threads.

----

The patch makes handling of SafeFetch faults independent on whether the crashing thread is a JavaThread (indeed, whether we have a current Thread at all). This had been the case for AIX and Linux ppc, s390 before, since we already fixed this issue for our platform, so we know this works.

I also hauled the SafeFetch handling out of the platform dependent part of the signal handler into the generic signal handler. This removes some duplicate coding.

To be consistent, I moved the SafeFetch handling for Zero up into the generic signal handler too. Zero did not have a problem, but this reduces code.

I added a gtest which reproduces the issue and used that to check that the patch works.

Thanks, Thomas

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8257828](https://bugs.openjdk.java.net/browse/JDK-8257828): SafeFetch may crash if invoked in non-JavaThreads

### Reviewers
 * [Martin Doerr](https://openjdk.java.net/census#mdoerr) (@TheRealMDoerr - **Reviewer**) ⚠️ Review applies to a7a94321e095dcc09b0e5580740da6b288d159f7
 * [Kim Barrett](https://openjdk.java.net/census#kbarrett) (@kimbarrett - **Reviewer**) ⚠️ Review applies to eb0efc1e4f75cefb6828edb1ec4643ff39390a80
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to a7a94321e095dcc09b0e5580740da6b288d159f7
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to eb0efc1e4f75cefb6828edb1ec4643ff39390a80

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1695/head:pull/1695`
`$ git checkout pull/1695`

### CODE REVIEW

**Review by TheRealMDoerr — APPROVED** [review]
Thank you for fixing it!
Good that we have signals_posix.cpp, now.
Seems like os::min_page_size() is an address which is not readable on any OS we have. So test looks also good to me.

**Review by kimbarrett — APPROVED** [review]
>  if (info != NULL && uc != NULL && thread != NULL) {

So the cause of the problem is the `thread != NULL` in that test. And just
removing all of those would fix the problem.

But I agree that eliminating all the (near) copied variants and merging this
functionality into signals_posix is good.

> This is unfortunate since SafeFetch is used [...]

Another place this hits is OopStorage, which uses SafeFetchN in two places.
Correct code shouldn't have a problem with one of them because the access
won't be invalid (if code is correct). But the other could be a problem,
though I think not seen in practice.

Might there be any similar problem on Windows?

This might be kind of related?
https://bugs.openjdk.java.net/browse/JDK-8185734

**Review by coleenp — APPROVED** [review]
This looks good to me.  Moving this to the posix signal handler is so nice.

**Review by dholmes-ora — APPROVED** [review]
Hi Thomas,
This looks good - I like the common handling (though I dislike the need for the zero case :( ).
Thanks,
David

**Comment by tstuefe** [comment]
Hi,

I made this work on Windows too (thanks @kimbarrett for reminding me). This involved fixing `HandleException`, which is 
roughly the pendant of ucontext_set_pc() for Windows. I also combined dumplicate coding for the three Windows architectures.

I excluded the `os_safefetch_negative` gtest for Windows, that won't work as long as we miss SEH for gtests, see https://bugs.openjdk.java.net/browse/JDK-8185734. This is not a big deal though since the `os_safefetch_negative_at_safepoint` gtest still works, which tests SafeFetch in the VMThread. I experimented with locally adding a SEH catcher around the SafeFetch call in the test, but I removed that coding because it introduced too much complexity for little gain.

I also fixed a small issue for AIX where the zero page is readable so we need another guaranteed-to-be-invalid address. Since we need invalid addresses in a number of places in text code, I plan to centralize this definition at some point.

Thanks, Thomas

**Review by kimbarrett — APPROVED** [review]
New Windows changes look okay to me, but I'm not an expert in that area.

File: src/hotspot/os/windows/os_windows.cpp [hunk 1]
```diff
@@ -2242,29 +2242,24 @@ int os::signal_wait() {
...
 #else
-  // Do not blow up if no thread info available.
-  if (thread != NULL) {
-    thread->as_Java_thread()->set_saved_exception_pc((address)(DWORD_PTR)exceptionInfo->ContextRecord->Eip);
+  #error unknown architecture
+#endif
+
+  // Save pc in thread
+  if (thread != nullptr && thread->is_Java_thread()) {
+    thread->as_Java_thread()->set_saved_exception_pc((address)(DWORD_PTR)exceptionInfo->ContextRecord->PC_NAME);
```
Comment (dholmes-ora) [inline src/hotspot/os/windows/os_windows.cpp:2258]: Is this a full fix? If the thread is not a JavaThread we no longer hit a problem in as_Java_thread() but is doing nothing for a non-JavaThread actually the right thing to do?

**Review by dholmes-ora — APPROVED** [review]
One query, but changes can go in as-is.
Thanks,
David

File: src/hotspot/os/windows/os_windows.cpp (hunk 1 above)
Comment (tstuefe) [reply src/hotspot/os/windows/os_windows.cpp:2258]: The saved_exception_pc mechanism is only implemented for JavaThreads. But all threads go through the code below, where we change the pc in the context structure.

**Comment by tstuefe** [comment]
> _Mailing list message from [David Holmes](mailto:david.holmes@oracle.com) on [hotspot-runtime-dev](mailto:hotspot-runtime-dev@openjdk.java.net):_
> 
> On 14/12/2020 3:24 pm, Thomas Stuefe wrote:
> 
> > On Mon, 14 Dec 2020 04:44:43 GMT, David Holmes <dholmes at openjdk.org> wrote:
> > > > Thomas Stuefe has updated the pull request incrementally with one additional commit since the last revision:
> > > > Make SafeFetch work on Windows + AIX fix
> > > 
> > > 
> > > src/hotspot/os/windows/os_windows.cpp line 2258:
> > > > 2256:   // Save pc in thread
> > > > 2257:   if (thread != nullptr && thread->is_Java_thread()) {
> > > > 2258:     thread->as_Java_thread()->set_saved_exception_pc((address)(DWORD_PTR)exceptionInfo->ContextRecord->PC_NAME);
> > > 
> > > 
> > > Is this a full fix? If the thread is not a JavaThread we no longer hit a problem in as_Java_thread() but is doing nothing for a non-JavaThread actually the right thing to do?
> > 
> > 
> > The saved_exception_pc mechanism is only implemented for JavaThreads. But all threads go through the code below, where we change the pc in the context structure.
> 
> My query is whether it makes sense for only JavaThreads to have this
> saved (C++?) exception pc? What role does it play and why would it not
> be applicable to all threads? (Sorry bit of a digression but trying to
> understand the bigger picture.)
> 
> Thanks,
> David

IIUC, JavaThread::_saved_exception_pc contains the pc from which a safepoint was triggered, and to which we may want to return once the safepoint was handled. Safepoint handling only works for java threads, so limiting this to JavaThread makes sense.

**Pull request closed without merging**